Boss controller update

cdgonz115 opened the pull request:
Just look at commit comments

cdgonz115 committed:
not finished

Added neccesary functionality to the bossmanger to make the scripts/actions/behaviors be able to run independent of the runFrames from the Phaseaction struct. Created a parent class for all actions in order to facilitate managing couroutines. Still having isse with making A0 repeat after every action that is not A0

cdgonz115 committed:
working as intended

Finally got A0 to repeat b4 every other RatAction, also fixed a bug in bbc that ended up only picking the first two elements of PossibleActions, also the variable actionQueue and the script PhaseReset seem unnecessary now, should probably delete them, also there is currently a bug with A3 where a ratRacer will spawn and not stay in the scene as it deoes not collide with any wall and essentially leave the ratboss stuck in A3 so we need to add a timer to them.

content-deleted commented on Assets/Scripts/Enemies/BossBehaviorController.cs:
```diff
@@ -4,108 +4,156 @@
 
-[RequireComponent (typeof(EnemyHealth))]
-public class BossBehaviorController : MonoBehaviour {
+public class BossAction : MonoBehaviour
```
Shouldnt this structure store a reference to an action?

cdgonz115 replied:
its an interface not a struct

content-deleted replied:
well its a class... not an interface (idk why i said structure instead of class because struct is an actual thing so saying (data) structure is confusing even if technically correct), interfaces cant describe what data is held in the classes which implement them, only give a guarantee of what methods exist on them etc

content-deleted commented on Assets/Scripts/Enemies/BossBehaviorController.cs:
```diff
@@ -4,108 +4,156 @@
         health = GetComponent<EnemyHealth>();
-        health.maxHealth = bossPhases.Select(x=> x.health).Sum();
-        health.onEnemyDamage+=checkHealth;
+        health.maxHealth = bossPhases.Select(x => x.health).Sum();
```
I think we should set up a linter for this project so that there aren't so many style changes in PRs
It makes it difficult to parse the actual changes in logic

content-deleted replied:
I started doing a bit of research on this and there are options available 
I'm not gonna lie and say I have time to set it up right now but I could help with hosting if we wanted to added code quality gates and a jenkins pipeline to the project

content-deleted commented on Assets/Scripts/Enemies/BossBehaviorController.cs:
```diff
@@ -4,108 +4,156 @@
+        if (bossPhases[currentPhase].startBehaviour != null)
+        {
+            bossPhases[currentPhase].startBehaviour.enabled = true;
+            while (bossPhases[currentPhase].startBehaviour.actionRunning)
```
honestly i would have had a much more complex solution but this works
we can have higher order controllers without having to adjust this one too

cdgonz115 replied:
yeah i thought this was the easiest way to handle a starting and  was planning to do the same for the ending action

content-deleted commented on Assets/Scripts/Enemies/BossBehaviorController.cs:
```diff
@@ -4,108 +4,156 @@
-            currentAction = bossPhases[currentPhase].PossibleActions.Where(x=> !x.Equals(currentAction)).ElementAt( rand.Next(bossPhases[currentPhase].PossibleActions.Count-1) );
+    public void ChangeAction()
+    {
+        foreach(PhaseAction p in actionQueue) print(p.behavior);
```
remove print statement when not needed, output is slow

cdgonz115 replied:
oops pretty sure this was left because i was testing

content-deleted commented on Assets/Scripts/Enemies/BossBehaviorController.cs:
```diff
@@ -4,108 +4,156 @@
+        foreach(PhaseAction p in actionQueue) print(p.behavior);
+        if (currentAction.behavior != null) currentAction.behavior.enabled = false;
+        if (bossPhases[currentPhase].repeatingBehaviour != null && currentAction.behavior != bossPhases[currentPhase].repeatingBehaviour) currentAction.behavior = bossPhases[currentPhase].repeatingBehaviour;
+        else if (actionQueue.Any()) { currentAction = actionQueue[0]; actionQueue.RemoveAt(0); }
```
this is an absolute mess, thats kinda my fault but 
god i cant parse it at all 
the diff makes it bad too
this needs to be spaced out more i dont know why i thought that should be on the same line

cdgonz115 replied:
lol idk either

cdgonz115 replied:
i added some comments to help clear it up

content-deleted commented on Assets/Scripts/Enemies/BossBehaviorController.cs:
```diff
@@ -4,108 +4,156 @@
+        if (bossPhases[currentPhase].repeatingBehaviour != null && currentAction.behavior != bossPhases[currentPhase].repeatingBehaviour) currentAction.behavior = bossPhases[currentPhase].repeatingBehaviour;
+        else if (actionQueue.Any()) { currentAction = actionQueue[0]; actionQueue.RemoveAt(0); }
+        else if (bossPhases[currentPhase].PossibleActions.Count > 1)
+            currentAction = bossPhases[currentPhase].PossibleActions.Where(x => !x.Equals(currentAction)).ElementAt(rand.Next(bossPhases[currentPhase].PossibleActions.Count));
```
Need to change this to evaluate with the change provided in the actions 
IE when the manager is started the numbers are statically evaluated to their percentages and replaced (so that the chance of all actions adds to 1) and then that number can be used probably, need to think about the math efficenty for selecting randomly from a list

content-deleted commented on Assets/Scripts/Enemies/BossBehaviorController.cs:
```diff
@@ -4,108 +4,156 @@
         else currentAction = bossPhases[currentPhase].PossibleActions.FirstOrDefault();
-
-        if(currentAction.behavior != null) StartCoroutine(startAction(currentAction, true));
+        if (currentAction.behavior != null) StartCoroutine(startAction(currentAction, true));
```
space these lines out from the current action selection logic beforhand

content-deleted commented on Assets/Scripts/Enemies/BossBehaviorController.cs:
```diff
@@ -4,108 +4,156 @@
-    public struct PhaseAction {
+    public struct PhaseAction
+    {
+        public bool independent;
```
what does this do?

cdgonz115 replied:
since alot of the ratboss actions timings where handled by themselves in their own scripts i decided to add an option to the PhaseAction struct that lets the manager know if it should run the behavior either by the runframes variable which is handled by the bbc or the behavior themselves handle that, if the behavior themselves handle the timings then the behavior would be "independent"

nicholasngoc replied:
Small suggestion to add a comment in here to describe this so all the other systems will for sure know what it is?

content-deleted commented on Assets/Scripts/Enemies/BossBehaviorController.cs:
```diff
@@ -4,108 +4,156 @@
-            if(currentPhase != startedPhase) {
-                action.behavior.enabled = false;
-                yield break;
+        if (action.independent || action.behavior == bossPhases[currentPhase].repeatingBehaviour)
```
oh is this for actions which have internal timers? 
that should probably be the default anyway

cdgonz115 replied:
yes lol

content-deleted commented on Assets/Scripts/Enemies/BossBehaviorController.cs:
```diff
@@ -4,108 +4,156 @@
+            }
+            action.behavior.enabled = false;
+        }
+        for (int i = 0; i < ((action.behavior == bossPhases[currentPhase].repeatingBehaviour) ? 0 : action.cooldownFrames); i++)
```
why is this necessary

cdgonz115 replied:
looking back i dont think it is, i think originally i had it out there but then decided to move it inside the if statement, still it be easiest to just put the action.behavior.enabled = false; after the if else statement

content-deleted commented on Assets/Scripts/Enemies/Ratboss/RatbossA0.cs:
```diff
@@ -4,10 +4,11 @@
 
-public class RatbossA0 : MonoBehaviour
-{ 
+public class RatbossA0 : BossAction
```
Okay so it does get inherited instead of MonoBehaviour
Please use `abstract` in the class definition to make that more clear

cdgonz115 replied:
will do

content-deleted commented on Assets/Scripts/Enemies/Ratboss/RatbossA3.cs:
```diff
@@ -88,34 +88,40 @@ void Awake()
-        yield return A0.shakedoors();
-
         leftSpawn = Random.Range(yPosLR - (yLengthLR / 2), yPosLR + (yLengthLR / 2));
         racerL.transform.position = new Vector3(leftWall.transform.position.x + (xLengthLR / 2)+0.5f, leftSpawn, -2);
```
Oh god I never saw any of this code jesus fucking christ how horrifying

content-deleted reviewed (changes requested):
If it works it works
There are a few things we can probably clean up here but I need to get the queue system actually working and make a few other changes

cdgonz115 committed:
updates

some updates based on Jacobs feedback, I added some tags to deal with a n issue with the rat racers where they were not getting destroying when they collided with the player or the wall, also added some comments and spaced out some lines to help facilitate understand the bbc, and added some very minor changes to it regarding repeating the same line multiple times when it was unnecessary

cdgonz115 committed:
more updates

Changed the BulletSpawner Script to inherit from BossAction instead of MonoBehaviour that way it could be used in the update manager, hid selectionChance variable as it is currently not working,  Changed the ChangePhase method to a coroutine to add the same functionality to the endBehaviour as the startBehaviour,
Fixed an unintended bug that would not allow the last possible action behavior to be picked due to the addition of the repeatingBehaviour mechanic. Also renabled the testBoss in the scene.

danielwaker commented:
good for now
